feat: custom retry alloy middleware - handling server errors

shunsukew opened the pull request:
Partially resolve https://github.com/alchemyplatform/rundler/issues/886

## Proposed Changes

Retry requests when having defined providers' server errors with the cost of latency increase.
Proposed approach is modified version of alloy `RetryBackoffLayer` https://github.com/alloy-rs/alloy/blob/main/crates/transport/src/layers/retry.rs, instead of implementing another custom middleware. This is to avoid unnecessary `ResponsePacket` decodings for efficiency.

shunsukew commented on crates/provider/src/alloy/provider_retry.rs:
```diff
@@ -0,0 +1,324 @@
+                    // try to extract the requested backoff from the error or compute the next
+                    // backoff based on retry count
+                    let backoff_hint = this.policy.backoff_hint(&err);
+                    let next_backoff = backoff_hint.unwrap_or_else(|| this.initial_backoff());
```
perhaps should better set exponential backoff when `backoff_hint` is None.

andysim3d commented on crates/provider/src/alloy/provider_retry.rs:
```diff
@@ -0,0 +1,324 @@
+use tokio::time::sleep;
+use tower::{Layer, Service};
+use tracing::trace;
+#[cfg(target_family = "wasm")]
```
we don't have plan to support rundler in wasm env, could you remove this, and only keep branch of not wasm?

andysim3d commented on crates/provider/src/alloy/provider_retry.rs:
```diff
@@ -0,0 +1,324 @@
+#[non_exhaustive]
+pub(crate) struct ProviderRetryPolicy;
+
+/// [RetryPolicy] defines logic for which [TransportError] instances should
```
could you use https://github.com/alloy-rs/alloy/blob/main/crates/transport/src/layers/retry.rs#L115-L121 instead of redefine the trait?

andysim3d commented on crates/provider/src/alloy/provider_retry.rs:
```diff
@@ -0,0 +1,324 @@
+        let inner = self.inner.clone();
+        let this = self.clone();
+        let mut inner = std::mem::replace(&mut self.inner, inner);
+        Box::pin(async move {
```
we use BoxFuture in rundler, you could find an example: https://github.com/alchemyplatform/rundler/blob/main/crates/provider/src/alloy/metrics.rs#L92

andysim3d reviewed (commented):
Thanks for the contribution! left some minor comments

shunsukew commented:
> Thanks for the contribution! left some minor comments

Thank you for the review! I changed to use `BoxFuture`.
It seems I added changes before you submit review comments, codes has changed since your last review, and now this PR contains separate `RetryBackoffLayer` for provider server errors.

This is for having different retry logics and configs from the one available in alloy middleware. Alloy one is solely focuses on rate limit `retry-able error (like resource exhausted)`, while custom layer from this PR focuses on retry-able server errors `inconsistent error(like unknown blocks)`.

Please let me know if prev version of middleware is more preferable. I can revert it back.

shunsukew commented:
> @shunsukew sorry for the long wait - your pr lgtm, just need you run cargo sort to make sure your cargo file is well sorted.

Thank you so much. sure, pushed commit after `cargo sort`

dancoombs commented on crates/provider/src/alloy/provider_error_retry.rs:
```diff
@@ -0,0 +1,151 @@
+
+static BLOCK_NOT_FOUND_RE: Lazy<Regex> = Lazy::new(|| {
+    // case-insensitive; 0x + 64 hex chars
+    Regex::new(r"(?i)^block 0x[0-9a-f]{64} not found$").unwrap()
```
In our recent testing we've also seen this error come up as "Unknown block" - would be great to add that to the regex here as well

shunsukew commented:
> Hey @shunsukew - we had an urgent need for this layer. (thanks for doing the upfront work!)
> 
> I refactored it a bit to look like our existing layer. And fixed a few things I found in testing on top of the alchemy RPC.
> 
> Going to close this in favor of: #1218 (cherry-picked your commits so you'll still have attribution 🙏 )

Sure, thank you so much!
